### PR TITLE
ore: add MetricsFutureExt which exposes Futures combinators to report to prometheus

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR creates a `MetricsFutureExt` which adds methods to types that implement `Future` for reporting timing data to prometheus. Specifically it adds the following functions:

* `fn wall_time(self)`
* `fn exec_time(self)`

Now users can write code like:
```
let result = my_future
  .inc_by(metrics.slow_queries.with_label_values(&[query_name]))
  .await;
```

Which will end up reporting the wall time of `Future` to the provided prometheus Counter. There is also the `observe(prometheus::Historgram)` method which can record to histograms.

### Motivation

* This PR adds a known-desirable feature.

Came up in [Slack](https://materializeinc.slack.com/archives/CMH6PG4CW/p1692217036046749) recently when I asked if we had any tools for reporting the timing of futures.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/ore/src/metrics.rs [hunk 1]
```diff
@@ -403,5 +410,369 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+impl<F> WallTimeFuture<F> {
+    /// Records the duration of this [`Future`] with the provided labels.
+    pub fn with_label_values(mut self, labels: &'static [&'static str]) -> Self {
+        self.labels = Some(labels);
+        self
+    }
+
+    /// Sets a minimum threshold. Any [`Future`]s whose duration is less than the provided
+    /// threshold will not get reported.
+    pub fn with_min_threshold(mut self, threshold: Duration) -> Self {
```
Comment (pH14) [inline src/ore/src/metrics.rs:508]: This feels overly specific to me... it seems like a relatively rare case to want to pre-filter which durations get recorded. For instance with a histogram, wouldn't all of the short-lived events get bucketed into some small bucket, so later analysis can exclude that one if needed?

If this is needed, I'd be partial to `with_filter(mut self, filter: Fn(Duration) -> bool)` so the caller can decide how to sample individual measurements

**Review by pH14 — COMMENTED** [review]
Very cool! This seem handy, thanks for typing this one up :D

File: src/ore/src/metrics.rs [hunk 2]
```diff
@@ -403,5 +410,369 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+        histogram: prometheus::HistogramVec,
+        unit: TimeUnit,
+    ) -> WallTimeFuture<F>;
+
+    /// Records the total amount of time for which a [`Future`] was executing.
+    ///
+    /// More specifically, every time the `Future` is polled it records how long that individual
+    /// call took, and maintains a running sum until the `Future` completes. Then we report that
+    /// duration to the provided histogram.
+    fn record_exec_time(
```
Comment (pH14) [inline src/ore/src/metrics.rs:441]: any early thoughts on when one would want `record_wall_time` vs `record_exec_time`? they both seem pretty interesting, and a large delta between them would also be revealing when understanding system behavior

Comment (pH14) [reply src/ore/src/metrics.rs:441]: might be useful to capture some recommendation in the doc

File: src/ore/src/metrics.rs [hunk 3]
```diff
@@ -403,5 +410,369 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+pub trait MetricsFutureExt<F> {
+    /// Records the amount of time it takes a [`Future`] to complete according to "the clock on
+    /// the wall".
+    ///
+    /// More specifically, it records the instant at which the `Future` was first polled, and the
+    /// instant at which the `Future` completes. Then reports the duration between those two
+    /// instances to the provided histogram.
+    fn record_wall_time(
+        self,
+        histogram: prometheus::HistogramVec,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/ore/src/metrics.rs]: Great suggestion! I re-worked the API to be generic over both, e.g.
```
my_future
  .wall_time()
  .observe(metric.slow_queries_hist)
  .await;
```
or
```
my_future
  .wall_time()
  .inc_by(metric.slow_queries_cnt)
  .await;
```

A type level flag prevents the user from calling just `.wall_time()` without also calling `.observe(...)` or `.inc_by(...)`

File: src/ore/src/metrics.rs [hunk 4]
```diff
@@ -403,5 +410,369 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+
+        // Only record the Duration if it's greather than our threshold.
+        let threshold = this.min_threshold.unwrap_or(Duration::from_micros(0));
+        if duration > threshold {
+            let duration = match this.unit {
+                TimeUnit::Milliseconds => f64::cast_lossy(duration.as_millis()),
+                TimeUnit::Seconds => duration.as_secs_f64(),
+            };
+            this.histogram
+                .with_label_values(this.labels.unwrap_or(&[]))
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/ore/src/metrics.rs]: Updated to make this change!

File: src/ore/src/metrics.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/ore/src/metrics.rs]: This makes sense, I changed the API to be `with_filter(...)`!

**Comment by ParkMyCar** [comment]
Thanks for all of the feedback @pH14! I'm much happier with the API now than the original attempt

File: src/ore/src/metrics.rs [hunk 5]
```diff
@@ -403,5 +408,481 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+
+    /// Records the total number of seconds for which a [`Future`] was executing.
+    ///
+    /// More specifically, every time the `Future` is polled it records how long that individual
+    /// call took, and maintains a running sum until the `Future` completes. Then we report that
+    /// duration to the provided metric.
+    ///
+    /// # Wall Time vs Execution Time
+    ///
+    /// There is also [`MetricsFutureExt::exec_time`], which measures how long a [`Future`] spent
```
Comment (pH14) [inline src/ore/src/metrics.rs:457]: nit: copypasta

File: src/ore/src/metrics.rs [hunk 6]
```diff
@@ -403,5 +408,481 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+            filter: self.filter,
+        }
+    }
+}
+
+impl<F, M> WallTimeFuture<F, M> {
+    /// Specifies a filter which much return `true` for the wall time to be recorded.
+    ///
+    /// This can be particularly useful if you have a high volume `Future`, e.g. Coordinator
+    /// messages, and you only want to record ones that take a long time to complete.
```
Comment (pH14) [inline src/ore/src/metrics.rs:575]: nit: let's keep `ore` free of references to MZ itself

File: src/ore/src/metrics.rs [hunk 7]
```diff
@@ -403,5 +408,481 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+    }
+}
+
+impl<F: Future, M: DurationMetric> Future for WallTimeFuture<F, M> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // Record when we started this Future.
```
Comment (pH14) [inline src/ore/src/metrics.rs:588]: nit: I think the inline comments here are self-explanatory with the code :)

File: src/ore/src/metrics.rs [hunk 8]
```diff
@@ -403,5 +408,481 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
...
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // Record the total amount of time this call to poll took.
+        let start = Instant::now();
+        let result = this.fut.poll(cx);
+        let duration = Instant::now().duration_since(start);
+
+        // Update our running sum.
```
Comment (pH14) [inline src/ore/src/metrics.rs:696]: likewise with these inline comments

**Review by pH14 — APPROVED** [review]
Left some small nits but otherwise LGTM! The way you used `UnspecifiedMetric` to force compile time errors without use is melting my mind, so cool

File: src/ore/src/metrics.rs (hunk 7 above)
Comment (ParkMyCar) [reply src/ore/src/metrics.rs]: Done!

File: src/ore/src/metrics.rs (hunk 6 above)
Comment (ParkMyCar) [reply src/ore/src/metrics.rs]: Makes sense!

**Comment by ParkMyCar** [comment]
Thanks for iterating on this with me @pH14!

**Pull request merged**